Docs/Book/Cookbook_v2.rst

```diff
@@ -0,0 +1,1010 @@
+| **Index ID#:** RDKitCB_1
+| **Summary:** Draw a molecule in black and white.
+
+.. testcode::
+
+   from rdkit import Chem
+   from rdkit.Chem.Draw import IPythonConsole
+   from rdkit.Chem import Draw
+
+.. testcode::
+
+   ms = [Chem.MolFromSmiles(x) for x in ('Cc1onc(-c2ccccc2)c1C(=O)N[C@@H]1C(=O)N2[C@@H](C(=O)O)C(C)(C)S[C@H]12','CC1(C)SC2C(NC(=O)Cc3ccccc3)C(=O)N2C1C(=O)O.[Na]')]
+   Draw.MolsToGridImage(ms)
+   
+.. image:: images/RDKitCB_1_im0.png
+
+.. testcode::
+
+   IPythonConsole.drawOptions.useBWAtomPalette()
+   Draw.MolsToGridImage(ms)
+
+.. image:: images/RDKitCB_1_im1.png
+
+Highlight a Substructure in a Molecule
+=======================================
+
+| **Author:** Greg Landrum
+| **Source:** `<https://gist.github.com/greglandrum/5d45b56afe75603b955103cdd0d8e038>`_
+| **Index ID#:** RDKitCB_2
+| **Summary:** Draw a molecule with a substructure highlight.
+
+.. testcode::
+
+   from rdkit import Chem
+   from rdkit.Chem.Draw import IPythonConsole
+
+.. testcode::
+
+   m = Chem.MolFromSmiles('c1cc(C(=O)O)c(OC(=O)C)cc1')
+   substructure = Chem.MolFromSmarts('C(=O)O')
+   print(m.GetSubstructMatches(substructure))
+
+.. testoutput::
+   
+   ((3, 4, 5), (8, 9, 7))
+   
+.. testcode::
+   
+   m
+
+.. image:: images/RDKitCB_2_im0.png
+   
+
+Without Implicit Hydrogens
+===========================
+
+| **Author:** Greg Landrum
+| **Source:** `<https://sourceforge.net/p/rdkit/mailman/message/36746387/>`_ and `<https://gist.github.com/greglandrum/9283aeadfb66d0fe8a2900e63fb10f3e>`_
+| **Index ID#:** RDKitCB_17
+| **Summary:** Draw a molecule without implicit hydrogens
+
+.. testcode::
+
+   from rdkit import Chem
+   from rdkit.Chem.Draw import IPythonConsole
+   m = Chem.MolFromSmiles('[Pt](Cl)(Cl)(N)N')
+   m
+
+.. image:: images/RDKitCB_17_im0.png
+
+.. testcode::
+
+   for atom in m.GetAtoms():
+       atom.SetProp("atomLabel", atom.GetSymbol())
+   m
+
+.. image:: images/RDKitCB_17_im1.png
+
+
+Rings, Aromaticity, and Kekulization
+*************************************
+
+Count Ring Systems
+=====================
+
+| **Author:** Greg Landrum
+| **Source:** `<https://gist.github.com/greglandrum/de1751a42b3cae54011041dd67ae7415>`_
+| **Index ID#:** RDKitCB_3
+| **Summary:** Count ring systems in a molecule
+
+.. testcode::
+
+   from rdkit import Chem
+   from rdkit.Chem.Draw import IPythonConsole
+
+.. testcode::
+
+   def GetRingSystems(mol, includeSpiro=False):
+        ri = mol.GetRingInfo()
+        systems = []
+        for ring in ri.AtomRings():
+            ringAts = set(ring)
+            nSystems = []
+            for system in systems:
+                nInCommon = len(ringAts.intersection(system)) 
+                if nInCommon and (includeSpiro or nInCommon>1):
+                    ringAts = ringAts.union(system)
+                else:
+                    nSystems.append(system)
+            nSystems.append(ringAts)
+            systems = nSystems
+        return systems
```

Comment:
This function first caught my eye because it seemed more confusing than it might need to be. But then I started to wonder if it was correct. I've asked Dan Nielschneider to take a look at it if he has time.

Reply:
yeah, it's actually not a trivial thing to do. I wasn't able to make it simpler than this

Reply:
I think this logically very tidy (compare to networkX's biconnected_components, code, for instance!). Maybe it looks confusing because it's replacing the systems variable in the loop. I wonder if doing something like the below would make that more explicit:

    def GetRingSystems(mol, includeSpiro=False):
        """Group rings into "systems" of connected rings"""

        def merge_into_systems(ringAts, systems, includeSpiro=False):
            """merge systems connected with ringAts into one system"""
            for system in systems:
                nInCommon = len(ringAts.intersection(system))
                if nInCommon and (includeSpiro or nInCommon > 1):
                    ringAts = ringAts.union(system)
                else:
                    yield system
            yield ringAts

        ri = mol.GetRingInfo()
        systems = []
        for ring in ri.AtomRings():
            ringAts = set(ring)
            systems = list(merge_into_systems(ringAts, systems, includeSpiro))
        return systems

Reply:
Thanks. This does look useful, but for some reason I can't get nested functions like this to pass doctest, even though it works great in IPython. Not sure why, so I did not add this in for now.